File: scripts/autopilot/pr-refs.py
#!/usr/bin/env python3
"""pr-refs.py — the ONE reference-detection predicate for autopilot open-PR
guards (issue #3852).

Reads `gh pr list --json headRefName,body` JSON from stdin and prints the
space-separated set of issue numbers that the open PRs reference, via EITHER:

  * the `issue-<N>-<slug>` head-branch convention (hydra-dev's branch name), OR
  * a keyword ref in the PR body — a GitHub closing verb (Closes / Fixes /
    Resolves, any tense) OR the non-closing `Refs #N` form.

This is the shared extractor issue #3852 asks for, and since #4334 it is the
ONLY copy: collect-state.sh computes all three of its in-flight exclusion
sets by piping its `gh pr list` payload through THIS script (no selector =
the union `ORCH_INFLIGHT_ISSUES`; `--source branch` / `--source body` = the
per-channel subsets its Candidate Exclusion telemetry attributes, #3964),
recover-stale.sh pipes its payload through the zero-arg union form, and
reap.py imports the narrower `closing_issues()` predicate. A change to the
reference-detection rule (a new closing verb, a branch-naming convention
change) is therefore made ONCE here.

`closing_issues()` (issue #4045) is a second, NARROWER predicate over the
same JSON shape: issue numbers an open PR actually CLOSES (body closing verb
only — never the branch-name convention, never the non-closing `Refs #N`
form). `reap.py` uses it to promote an issue from `ready-for-agent` to
`needs-qa` once a real closing PR exists, which is a stricter bar than
`referenced_issues()`'s "this PR is at least related to the issue".

`branch_issues()` / `bodyref_issues()` (issue #4334) expose the two evidence
CHANNELS of `referenced_issues()` separately, for the per-source attribution
collect-state.sh's Candidate Exclusion telemetry needs — which matcher
actually fired for a given anchor. They are strict subsets of the union by
construction (same regexes, one channel each).

Pure: stdin JSON in, stdout numbers out. It NEVER shells out to `gh` — the
callers (collect-state.sh, recover-stale.sh) own the `gh pr list` call and
the never-abort degradation contract. Any parse error prints nothing and
exits 0: an empty result is the caller's "no open PR" signal, which falls
through to today's behaviour (re-queue to ready-for-agent). An unknown
`--source` value is the one loud failure (exit 2): the bash call sites wrap
the invocation in `2>/dev/null || true`, so it degrades to the documented
empty-set no-op while staying diagnosable when run by hand.
"""

import json
import re
import sys

# `issue-<N>` head-branch prefix (the `-<slug>` tail is optional). `\b` after
# the digits stops `issue-385` matching branch `issue-3852-foo`. `re.match`
# anchors at the start of the head ref.
_BRANCH_RE = re.compile(r"issue-(\d+)\b")

# PR-body keyword refs: GitHub closing verbs (close / fix / resolve, any tense)
# PLUS the non-closing `Ref(s) #N` form (issue #3851). Case-insensitive.
_BODY_RE = re.compile(
    r"\b(?:close[sd]?|fix(?:e[sd])?|resolve[sd]?|refs?)\s*:?\s+#(\d+)\b",
    re.IGNORECASE,
)

# Closing-verb-only subset of _BODY_RE (issue #4045) — deliberately EXCLUDES
# the non-closing `Ref(s) #N` form and never matches on branch name alone.
# GitHub's own auto-close mechanism keys on exactly these verbs in a PR body,
# so this is the predicate for "this PR marks the issue done", not merely
# "this PR is related to the issue".
_CLOSE_RE = re.compile(
    r"\b(?:close[sd]?|fix(?:e[sd])?|resolve[sd]?)\s*:?\s+#(\d+)\b",
    re.IGNORECASE,
)

def _prs(pr_json):
    """Parse a `gh pr list --json` payload into its PR dict rows.

    Anything unparsable — empty stdin (a failed `gh pr list`), non-JSON, a
    non-list top level — degrades to NO rows, which every selector then folds
    into the empty set: the fail-open contract the never-abort callers rely on.
    """
    try:
        prs = json.loads(pr_json)
    except Exception:
        return []
    if not isinstance(prs, list):
        return []
    return [p for p in prs if isinstance(p, dict)]


def _branch_refs(pr):
    """The issue numbers one PR row references via its head-branch name."""
    m = _BRANCH_RE.match(pr.get("headRefName") or "")
    return {int(m.group(1))} if m else set()


def _body_refs(pr):
    """The issue numbers one PR row references via body keyword refs."""
    return {int(m.group(1)) for m in _BODY_RE.finditer(pr.get("body") or "")}


def referenced_issues(pr_json):
    """Return the set of ints referenced by the open-PR JSON on stdin.

    The UNION of both evidence channels: the `issue-<N>` head-branch
    convention OR a body keyword ref (closing verb or `Refs #N`).
    """
    out = set()
    for pr in _prs(pr_json):
        out |= _branch_refs(pr)
        out |= _body_refs(pr)
    return out


def branch_issues(pr_json):
    """Return the set of ints referenced via the head-branch convention only.

    One channel of `referenced_issues()` in isolation (issue #4334), exposed
    for collect-state.sh's per-source Candidate Exclusion telemetry (#3964):
    distinguishing "the branch matcher fired" from "the body matcher fired".
    """
    out = set()
    for pr in _prs(pr_json):
        out |= _branch_refs(pr)
    return out


def bodyref_issues(pr_json):
    """Return the set of ints referenced via PR-body keyword refs only.

    The other channel of `referenced_issues()` in isolation (issue #4334),
    same telemetry rationale as `branch_issues()`.
    """
    out = set()
    for pr in _prs(pr_json):
        out |= _body_refs(pr)
    return out


def closing_issues(pr_json):
    """Return the set of ints an open PR ACTUALLY CLOSES (issue #4045).

    Narrower than `referenced_issues()` on purpose: a bare `issue-<N>`
    branch-name match or a non-closing `Refs #N` body keyword both count as
    "referenced" (enough to say a dev_orch dispatch didn't stall — see
    `reap.py`'s `_dev_orch_pr_exists_for_anchor`), but neither means the PR
    is actually done and ready for review. Only a real GitHub closing verb
    in the PR body (Closes/Fixes/Resolves, any tense) counts here — this is
    the predicate `reap.py` uses to promote an issue from `ready-for-agent`
    to `needs-qa`, and promoting on a merely-related PR would move an issue
    into the review lane before it's reviewable.
    """
    out = set()
    for pr in _prs(pr_json):
        for m in _CLOSE_RE.finditer(pr.get("body") or ""):
            out.add(int(m.group(1)))
    return out


def _selector_for(argv):
    """Map argv onto a predicate. Zero args = the union (the contract
    recover-stale.sh and the hydra-dev parent flow already depend on);
    `--source branch|body` picks one channel; anything else exits 2."""
    if not argv:
        return referenced_issues
    if len(argv) == 2 and argv[0] == "--source" and argv[1] in ("branch", "body"):
        return branch_issues if argv[1] == "branch" else bodyref_issues
    sys.stderr.write(
        "usage: pr-refs.py [--source branch|body] < gh-pr-list-JSON\n"
        f"unknown arguments: {' '.join(argv)}\n"
    )
    sys.exit(2)


def main():
    nums = _selector_for(sys.argv[1:])(sys.stdin.read())
    if nums:
        # Space-separated, sorted for determinism. No trailing newline needed —
        # callers word-split. An empty set prints nothing (= "no open PR").
        sys.stdout.write(" ".join(str(n) for n in sorted(nums)))


if __name__ == "__main__":
    main()

File: src/github/pr-refs.ts
/**
 * src/github/pr-refs.ts — the PR-ref detection predicate, ported from
 * `scripts/autopilot/pr-refs.py` (ADR-0040 Decision 4 row 6, Decision 5;
 * wayfinder #4517, issue #4683).
 *
 * `pr-refs.py` is the ONE reference-detection predicate the bash autopilot
 * lane (`collect-state.sh`, `recover-stale.sh`) and `reap.py` share — see its
 * own docstring for the full history. This module ports the same three
 * regexes and three matcher semantics into TypeScript for the two
 * `scripts/ci/*.ts` consumers (`epic-close.ts`, `design-concept-reconcile-check.ts`)
 * that previously hand-rolled their own copy of the closing-verb alternation.
 * `pr-refs.py` itself is NOT edited or replaced — it stays the bash-facing
 * predicate until #4686/#4688 give the bash lane a TS-backed path. This
 * module and the Python script are two independent implementations kept in
 * sync by a source-string parity test (`test/github-pr-refs.test.mts`, the
 * #3965 convention already used for `STRICT_BLOCKER_PATTERN_SOURCES` vs the
 * `collect-state.sh` jq literal): a change to either side alone fails the
 * test.
 *
 * # Family membership (CONTEXT.md: GitHub CLI Adapter)
 *
 * This module joins the `src/github/*` family as a PURE predicate sibling of
 * `src/github/blockers.ts` — no `gh`/`git` spawn, no import of
 * `src/github/exec.ts`. It is a leaf: stdin-shaped rows in, `ReadonlySet<number>`
 * out, nothing else.
 *
 * # Purity contract
 *
 * No `node:fs` / `node:child_process` / `fetch` / `process.env`, no
 * module-scope side effects beyond regex construction, no classes. Every
 * export is a pure function or a constant. No function throws on a row with
 * null/undefined/missing `headRefName`, `body`, or `title` — such a row
 * simply contributes nothing to the result.
 */

/**
 * The GitHub closing-verb alternation — close/fix/resolve, any tense — as a
 * bare regex-alternation fragment (no `\b`, no anchors, no flags). This is
 * the SINGLE TypeScript home for the verb list: `BODY_RE` and `CLOSE_RE`
 * below compose it, and the two `scripts/ci/*.ts` consumers import this
 * constant and compose their OWN existing tail (differing at the edges —
 * `\b` placement, `\s*` vs `\s+`) so each file's resulting `.source` stays
 * byte-identical to the inline literal it replaces.
 */
export const CLOSING_VERB_ALTERNATION = "close[sd]?|fix(?:e[sd])?|resolve[sd]?";

/**
 * `issue-<N>` head-branch prefix (hydra-dev's branch-naming convention; the
 * `-<slug>` tail is optional). `\b` after the digits stops `issue-385`
 * matching branch `issue-3852-foo`. No flags — mirrors `pr-refs.py`'s
 * `_BRANCH_RE`, which anchors via `re.match` (start-of-string) rather than a
 * `^` in the pattern source; {@link branchRef} reproduces that anchoring by
 * requiring `match.index === 0`.
 */
export const BRANCH_RE = /issue-(\d+)\b/;

/**
 * PR-body keyword refs: a GitHub closing verb OR the non-closing `Ref(s) #N`
 * form. Case-insensitive, global (consumed only via `matchAll`, never an
 * `.exec` loop, so the constant's `lastIndex` never advances). Mirrors
 * `pr-refs.py`'s `_BODY_RE`.
 */
export const BODY_RE = new RegExp(
  String.raw`\b(?:${CLOSING_VERB_ALTERNATION}|refs?)\s*:?\s+#(\d+)\b`,
  "gi",
);

/**
 * Closing-verb-only subset of {@link BODY_RE} — deliberately excludes the
 * non-closing `Ref(s) #N` form and never matches on branch name alone.
 * Mirrors `pr-refs.py`'s `_CLOSE_RE`.
 */
export const CLOSE_RE = new RegExp(
  String.raw`\b(?:${CLOSING_VERB_ALTERNATION})\s*:?\s+#(\d+)\b`,
  "gi",
);

/**
 * A `(#N)` anchor in a PR title — the `(#4236)`/`(#4130)` shape the GLM
 * drainer's `issue_has_merged_pr` jq rule also recognises. Global (consumed
 * only via `matchAll`).
 */
export const TITLE_ANCHOR_RE = /\(#(\d+)\)/g;

/**
 * The structural subset of a PR row this module needs. `PrRow` from
 * `src/github/prs.ts` satisfies this unchanged (its `body` field, when
 * absent, is simply not present — every field here is optional). No new
 * adapter type, no JSON parsing: parsing stays at the seam (`parsePrRows`).
 */
export interface PrRefRow {
  headRefName?: string | null;
  body?: string | null;
  title?: string | null;
}

/** The issue number one row references via its head-branch name, or `null`. */
function branchRef(row: PrRefRow): number | null {
  const headRefName = row.headRefName ?? "";
  const m = BRANCH_RE.exec(headRefName);
  if (!m || m.index !== 0) return null;
  const n = Number(m[1]);
  return Number.isFinite(n) && n > 0 ? n : null;
}

/** Every positive-integer capture group 1 match of a global `re` over `text`. */
function matchAllNumbers(re: RegExp, text: string): number[] {
  const out: number[] = [];
  for (const m of text.matchAll(re)) {
    const n = Number(m[1]);
    if (Number.isFinite(n) && n > 0) out.push(n);
  }
  return out;
}

/**
 * The union of both evidence channels: the `issue-<N>` head-branch
 * convention OR a body keyword ref (closing verb or `Refs #N`). Mirrors
 * `pr-refs.py`'s `referenced_issues()`.
 */
export function referencedIssues(prs: readonly PrRefRow[]): ReadonlySet<number> {
  const out = new Set<number>();
  for (const pr of prs) {
    const b = branchRef(pr);
    if (b !== null) out.add(b);
    for (const n of matchAllNumbers(BODY_RE, pr.body ?? "")) out.add(n);
  }
  return out;
}

/**
 * Issue numbers an open PR ACTUALLY CLOSES — `CLOSE_RE` over the body ONLY
 * (never the branch name, never `Refs #N`). Mirrors `pr-refs.py`'s
 * `closing_issues()`.
 */
export function closedIssues(prs: readonly PrRefRow[]): ReadonlySet<number> {
  const out = new Set<number>();
  for (const pr of prs) {
    for (const n of matchAllNumbers(CLOSE_RE, pr.body ?? "")) out.add(n);
  }
  return out;
}

/**
 * Issue numbers a (typically MERGED) PR references via a closing verb over
 * `title + body`, UNION a `(#N)` title anchor. This is the drainer's
 * `issue_has_merged_pr` rule (ADR-0040 Decision 4 row 7), consumed later by
 * #4686/#4687/#4690.
 */
export function mergedPrReferences(prs: readonly PrRefRow[]): ReadonlySet<number> {
  const out = new Set<number>();
  for (const pr of prs) {
    const combined = `${pr.title ?? ""}\n${pr.body ?? ""}`;
    for (const n of matchAllNumbers(CLOSE_RE, combined)) out.add(n);
    for (const n of matchAllNumbers(TITLE_ANCHOR_RE, pr.title ?? "")) out.add(n);
  }
  return out;
}
